Make fewer PositioningContexts when descending

mrobinson opened the pull request:
When descending and we have the option, don't create new
PositioningContexts just to update the static position of laid out
abspos descendants. Instead, use the new PositioningContextLength to
only update the newly added hoisted abspos boxes.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because they do not change behavior.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson commented:
@bors-servo try=wpt-2020

Loirooriol commented on components/layout_2020/positioned.rs:
```diff
@@ -175,20 +190,22 @@ impl PositioningContext {
     pub(crate) fn adjust_static_position_of_hoisted_fragments(
         &mut self,
         parent_fragment: &Fragment,
+        index: PositioningContextLength,
```
Nit: The comment above says

> This function is used to update that static position at every level of the fragment tree

But this can now be limited with `index`, so the comment may need some update.

mrobinson replied:
Okay. I've updated this comment.

Loirooriol reviewed (approved):
LGTM with nit.

But this is on top of #30057, which ended up having more failures than expected. So maybe you should land this 1st, only taking the necessary parts from #30057

mrobinson committed:
Make fewer PositioningContexts when descending

When descending and we have the option, don't create new
PositioningContexts just to update the static position of laid out
abspos descendants. Instead, use the new PositioningContextLength to
only update the newly added hoisted abspos boxes.

mrobinson commented:
Thanks for the review! I've gone ahead and taken the bits of #30057 that were relevant to this change.
